File: control/pid.py
"""control/pid.py -- PID controller for pan and tilt axes with anti-windup and derivative filtering.

Gain constants are defined as named module-level constants.
"""

from __future__ import annotations

from dataclasses import dataclass
import math

__all__ = [
    "DEFAULT_KP_PAN",
    "DEFAULT_KI_PAN",
    "DEFAULT_KD_PAN",
    "DEFAULT_KP_TILT",
    "DEFAULT_KI_TILT",
    "DEFAULT_KD_TILT",
    "DEFAULT_INTEGRAL_MAX",
    "DEFAULT_MAX_OUTPUT_STEP",
    "PID1D",
    "PanTiltPID",
]

# Named gain constants (units: rad/px for Kp, rad/(px*s) for Ki, rad*s/px for Kd)
# 1 pixel ~ 0.00078 rad for 640px @ 0.5 rad FOV.
DEFAULT_KP_PAN: float = 0.0003
DEFAULT_KI_PAN: float = 0.00001
DEFAULT_KD_PAN: float = 0.00001

DEFAULT_KP_TILT: float = 0.0003
DEFAULT_KI_TILT: float = 0.00001
DEFAULT_KD_TILT: float = 0.00001

# Anti-windup limit for integrated error term (units: px*seconds)
DEFAULT_INTEGRAL_MAX: float = 100.0

# Maximum single-step output delta limit (units: radians per frame step)
# 0.003 rad ~ 3.84 pixels per step max gimbal velocity limit (~6.8 deg/s)
DEFAULT_MAX_OUTPUT_STEP: float = 0.003


@dataclass
class PID1D:
    """Single-axis PID controller with integral clamping (anti-windup) and derivative filtering."""

    kp: float
    ki: float
    kd: float
    integral_limit: float = DEFAULT_INTEGRAL_MAX
    max_output_step: float = DEFAULT_MAX_OUTPUT_STEP
    d_filter_alpha: float = 0.25  # Low-pass filter coefficient for derivative

    _integral: float = 0.0
    _prev_error: float | None = None
    _d_filtered: float = 0.0

    def reset(self) -> None:
        """Reset integral accumulation, derivative history, and filter state."""
        self._integral = 0.0
        self._prev_error = None
        self._d_filtered = 0.0

    def update(self, error: float, dt: float, clamped: bool = False) -> float:
        """Compute PID output step for a given error and timestep dt (seconds).

        If clamped is True (e.g. joint hit saturation limit), integral accumulation
        is frozen to prevent windup.
        """
        if dt <= 0.0 or not math.isfinite(error):
            return 0.0

        # Derivative calculation with exponential moving average low-pass filter
        if self._prev_error is not None:
            raw_d = (error - self._prev_error) / dt
            self._d_filtered = self.d_filter_alpha * raw_d + (1.0 - self.d_filter_alpha) * self._d_filtered
        else:
            self._d_filtered = 0.0
        self._prev_error = error

        # Integral accumulation with anti-windup clamping
        if not clamped:
            self._integral += error * dt
            if self.integral_limit > 0.0:
                self._integral = max(-self.integral_limit, min(self.integral_limit, self._integral))

        # PID terms sum
        p_term = self.kp * error
        i_term = self.ki * self._integral
        d_term = self.kd * self._d_filtered

        output = p_term + i_term + d_term

        # Output saturation limit
        if self.max_output_step > 0.0:
            output = max(-self.max_output_step, min(self.max_output_step, output))

        return output


class PanTiltPID:
    """Combined 2-axis PID controller for pan and tilt axes."""

    def __init__(
        self,
        kp_pan: float = DEFAULT_KP_PAN,
        ki_pan: float = DEFAULT_KI_PAN,
        kd_pan: float = DEFAULT_KD_PAN,
        kp_tilt: float = DEFAULT_KP_TILT,
        ki_tilt: float = DEFAULT_KI_TILT,
        kd_tilt: float = DEFAULT_KD_TILT,
        integral_limit: float = DEFAULT_INTEGRAL_MAX,
        max_output_step: float = DEFAULT_MAX_OUTPUT_STEP,
    ) -> None:
        self.pan_pid = PID1D(
            kp=kp_pan,
            ki=ki_pan,
            kd=kd_pan,
            integral_limit=integral_limit,
            max_output_step=max_output_step,
        )
        self.tilt_pid = PID1D(
            kp=kp_tilt,
            ki=ki_tilt,
            kd=kd_tilt,
            integral_limit=integral_limit,
            max_output_step=max_output_step,
        )

    def reset(self) -> None:
        """Reset both pan and tilt PID controllers."""
        self.pan_pid.reset()
        self.tilt_pid.reset()

    def update(
        self,
        error_px: tuple[float, float],
        dt: float,
        pan_clamped: bool = False,
        tilt_clamped: bool = False,
    ) -> tuple[float, float]:
        """Compute (delta_pan_rad, delta_tilt_rad) from TrackResult.error_px.

        Sign convention:
        - error_px.x > 0 (target right of boresight) -> delta_pan > 0 (rotate right)
        - error_px.y < 0 (target above boresight) -> delta_tilt > 0 (rotate up)
        """
        err_x, err_y = error_px
        delta_pan = +self.pan_pid.update(err_x, dt, clamped=pan_clamped)
        delta_tilt = -self.tilt_pid.update(err_y, dt, clamped=tilt_clamped)
        return delta_pan, delta_tilt

File: src/control/pid.ts
/**
 * control/pid.ts - 1D and Pan-Tilt PID controllers with anti-windup and derivative filtering.
 * Faithful port of control/pid.py.
 */

import { PIDGains } from '../types/contracts';

export const DEFAULT_KP = 0.0003;
export const DEFAULT_KI = 0.00005;
export const DEFAULT_KD = 0.00002;
export const DEFAULT_MAX_OUTPUT_STEP = 0.003; // rad/step slew-rate limit (~3.8 px/step)
export const DERIVATIVE_ALPHA = 0.25; // Low-pass filter for derivative to suppress kick

export class PID1D {
  private kp: number;
  private ki: number;
  private kd: number;
  private maxStep: number;
  private integral: number = 0;
  private prevError: number | null = null;
  private filteredDerivative: number = 0;
  private integralLimit: number = 0.05;

  constructor(kp = DEFAULT_KP, ki = DEFAULT_KI, kd = DEFAULT_KD, maxStep = DEFAULT_MAX_OUTPUT_STEP) {
    this.kp = kp;
    this.ki = ki;
    this.kd = kd;
    this.maxStep = maxStep;
  }

  public setGains(gains: Partial<PIDGains>): void {
    if (gains.kp !== undefined) this.kp = gains.kp;
    if (gains.ki !== undefined) this.ki = gains.ki;
    if (gains.kd !== undefined) this.kd = gains.kd;
    if (gains.max_output_step !== undefined) this.maxStep = gains.max_output_step;
  }

  public reset(): void {
    this.integral = 0;
    this.prevError = null;
    this.filteredDerivative = 0;
  }

  public update(error: number, dt: number, isClamped: boolean = false): number {
    if (dt <= 0) return 0;

    // Proportional term
    const pTerm = this.kp * error;

    // Integral term with anti-windup: do not accumulate if joint was clamped in direction of error
    if (!isClamped) {
      this.integral += error * dt;
      this.integral = Math.max(-this.integralLimit, Math.min(this.integralLimit, this.integral));
    }
    const iTerm = this.ki * this.integral;

    // Derivative term with low-pass filtering
    let dTerm = 0;
    if (this.prevError !== null) {
      const rawD = (error - this.prevError) / dt;
      this.filteredDerivative = DERIVATIVE_ALPHA * rawD + (1 - DERIVATIVE_ALPHA) * this.filteredDerivative;
      dTerm = this.kd * this.filteredDerivative;
    }
    this.prevError = error;

    const rawOutput = pTerm + iTerm + dTerm;

    // Slew-rate limiting per timestep
    return Math.max(-this.maxStep, Math.min(this.maxStep, rawOutput));
  }
}

export class PanTiltPID {
  private panPID: PID1D;
  private tiltPID: PID1D;

  constructor(gains?: PIDGains) {
    const kp = gains?.kp ?? DEFAULT_KP;
    const ki = gains?.ki ?? DEFAULT_KI;
    const kd = gains?.kd ?? DEFAULT_KD;
    const maxStep = gains?.max_output_step ?? DEFAULT_MAX_OUTPUT_STEP;

    this.panPID = new PID1D(kp, ki, kd, maxStep);
    this.tiltPID = new PID1D(kp, ki, kd, maxStep);
  }

  public setGains(gains: PIDGains): void {
    this.panPID.setGains(gains);
    this.tiltPID.setGains(gains);
  }

  public reset(): void {
    this.panPID.reset();
    this.tiltPID.reset();
  }

  /**
   * Update PID:
   * Pan delta = +PID(error_x)
   * Tilt delta = -PID(error_y) (because image row y increases downward while elevation increases upward)
   */
  public update(
    errorPx: [number, number],
    dt: number,
    panClamped: boolean = false,
    tiltClamped: boolean = false
  ): [number, number] {
    const deltaPan = this.panPID.update(errorPx[0], dt, panClamped);
    const deltaTilt = -this.tiltPID.update(errorPx[1], dt, tiltClamped);
    return [deltaPan, deltaTilt];
  }
}
